diff --git a/NEWS b/NEWS
index 61b5d42f6..79f5b0e99 100644
--- a/NEWS
+++ b/NEWS
@@ -47,6 +47,9 @@ GNU coreutils NEWS                                    -*- outline -*-
 
 ** Improvements
 
+  df now recognizes these file systems as remote:
+  acfs, coda, fhgfs, gpfs, ibrix, ocfs2, and vxfs.
+
   stat and tail now know about the "devmem", "vboxsf", and "zonefs"
   file system types.  stat -f -c%T now reports the file system type,
   and tail -f uses polling for "vboxsf" and inotify for the others.
