From 1e764c5b9a3eb9a73370dbcbc04c2462c31aaf45 Mon Sep 17 00:00:00 2001
From: openEuler Buildteam <buildteam@openeuler.org>
Date: Wed, 29 Jul 2020 11:55:45 +0800
Subject: [PATCH] bugfix selinux flask

---
 m4/gnulib-comp.m4 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/m4/gnulib-comp.m4 b/m4/gnulib-comp.m4
index dead90e..0abf0bd 100644
--- a/m4/gnulib-comp.m4
+++ b/m4/gnulib-comp.m4
@@ -1860,10 +1860,10 @@ AC_DEFUN([gl_INIT],
     AC_LIBOBJ([select])
   fi
   gl_SYS_SELECT_MODULE_INDICATOR([select])
-  AC_CHECK_HEADERS([selinux/flask.h])
   gl_HEADERS_SELINUX_SELINUX_H
   gl_HEADERS_SELINUX_CONTEXT_H
   if test "$with_selinux" != no && test "$ac_cv_header_selinux_selinux_h" = yes; then
+    AC_CHECK_HEADERS([selinux/flask.h])
     AC_LIBOBJ([getfilecon])
   fi
   gl_SERVENT
-- 
2.23.0

