diff --git a/src/sort.c b/src/sort.c
index 0163ba18a..8671ea767 100644
--- a/src/sort.c
+++ b/src/sort.c
@@ -1651,7 +1651,7 @@ begfield (struct line const *line, struct keyfield const *key)
 /* Return the limit of (a pointer to the first character after) the field
    in LINE specified by KEY. */
 
-static char *
+static char * _GL_ATTRIBUTE_PURE
 limfield (struct line const *line, struct keyfield const *key)
 {
   char *ptr = line->text, *lim = ptr + line->length - 1;
